src/backend/cdb/motion/cdbmotion.c

```diff
@@ -1274,3 +1280,115 @@ UpdateSentRecordCache(int32 *sent_record_typmod)
 		*sent_record_typmod = NextRecordTypmod;
 	}
 }
+
+static int
+GetICTypeByName(const char *type_name)
+{
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (!pg_strcasecmp(IPCLayerImpls[i]->type_name, type_name))
+			return IPCLayerImpls[i]->ic_type;
+	}
+
+	return -1;
+}
+
+/*
+ * try best to find the most appropriate IPC layer implement.
+ */
+static int
+TryFindICType()
+{
+	/* IPCLayerImpls[0]->ic_type is valid at least */
+	Assert(CurrentIPCLayerImplNum);
+
+	int candidate_types_order[IPCLAYER_CANDIDATE_NUM] =
+		{Gp_interconnect_type, INTERCONNECT_TYPE_UDPIFC, IPCLayerImpls[0]->ic_type};
+
+	for (int i = 0; i < IPCLAYER_CANDIDATE_NUM; ++i)
+	{
+		for (int j = 0; j < CurrentIPCLayerImplNum; ++j)
+		{
+			/*
+			 * find candidate ic type in IPCLayerImpls and return it.
+			 */
+			if (candidate_types_order[i] == IPCLayerImpls[j]->ic_type)
+				return candidate_types_order[i];
+		}
+	}
+
+	/* never run here, just keep compiler slience */
+	Assert(0);
+	return INTERCONNECT_TYPE_UDPIFC;
+}
+
+void
+SetCurrentMotionIPCLayer(const char *type_name)
+{
+	int type;
+
+	/* do nothing before interconnect.so loaded. */
+	if (!process_shared_preload_libraries_done)
+		return;
+
+	if (type_name == NULL)
+		type_name = "";
+
+	type = GetICTypeByName(type_name);
+	if (type < 0)
+	{
+		if (IsUnderPostmaster)
+		{
+			/* backend process */
+			ereport(ERROR,
+					(errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+					 errmsg("No IPC layer implement found with type: \"%s\".",
+							type_name)));
+		}
+
+		/* postmaster process */
+		ereport(WARNING,
+				(errcode(ERRCODE_WARNING_GP_INTERCONNECTION),
+				 errmsg("No IPC layer implement found with type: \"%s\", "
+						"choose one from the loaded implements.",
+						type_name)));
+
+		type = TryFindICType();
+	}
+
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (IPCLayerImpls[i]->ic_type == type)
+		{
+			CurrentMotionIPCLayer = IPCLayerImpls[i];
+			Gp_interconnect_type  = CurrentMotionIPCLayer->ic_type;
+			break;
+		}
+	}
+
+	return;
+}
+
+void
+RegisterIPCLayerImpl(MotionIPCLayer *impl)
+{
+	if (CurrentIPCLayerImplNum >= MAX_NUMBER_TYPES)
+	{
+		ereport(WARNING,
```

Comment:
If the array is full, the user may not be aware of. How about to raise an error?

Reply:
fix with [commit](https://github.com/cloudberrydb/cloudberrydb/commit/8e8c769beb5dcf4fe8456ea299350620201e5275)